morpheus/stages/input/file_source.py

```diff
@@ -0,0 +1,319 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""File source stage."""
+
+import logging
+import time
+import typing
+from functools import partial
+from urllib.parse import urlsplit
+
+import fsspec
+import mrc
+from mrc.core import operators as ops
+
+from morpheus.cli import register_stage
+from morpheus.common import FileTypes
+from morpheus.config import Config
+from morpheus.config import PipelineModes
+from morpheus.io.deserializers import read_file_to_df
+from morpheus.messages import MessageMeta
+from morpheus.pipeline.preallocator_mixin import PreallocatorMixin
+from morpheus.pipeline.single_output_source import SingleOutputSource
+from morpheus.pipeline.stream_pair import StreamPair
+from morpheus.utils.directory_watcher import DirectoryWatcher
+
+logger = logging.getLogger(__name__)
+
+
+@register_stage("file-source", modes=[PipelineModes.FIL, PipelineModes.NLP, PipelineModes.OTHER])
+class FileSource(PreallocatorMixin, SingleOutputSource):
+    """
+    Load messages from a file.
+
+    Source stage is used to load messages from a file and dumping the contents into the pipeline immediately. Useful for
+    testing performance and accuracy of a pipeline.
+
+    Parameters
+    ----------
+    config : `morpheus.config.Config`
+        Pipeline configuration instance.
+    files : List[str]
+        List of paths to be read from, can be a list of S3 URLs (`s3://path`) and can include wildcard characters `*`
+        as defined by `fsspec`:
+        https://filesystem-spec.readthedocs.io/en/latest/api.html?highlight=open_files#fsspec.open_files
+    watch : bool, default = False
+        When True, will check `files` for new files and emit them as they appear.  (Note: `watch_interval` is
+        applicable when `watch` is True and there are no remote paths in `files`.)
+    watch_interval : float, default = 1.0
+        When `watch` is True, this is the time in seconds between polling the paths in `files` for new files.
+        (Note: Applicable when path in `files` are remote and when `watch` is True)
+    sort_glob : bool, default = False
+        If true, the list of files matching `input_glob` will be processed in sorted order.
+        (Note: Applicable when all paths in `files` are local.)
+    recursive : bool, default = True
+        If true, events will be emitted for the files in subdirectories matching `input_glob`.
+        (Note: Applicable when all paths in `files` are local.)
+    queue_max_size : int, default = 128
+        Maximum queue size to hold the file paths to be processed that match `input_glob`.
+        (Note: Applicable when all paths in `files` are local.)
+    batch_timeout : float, default = 5.0
+        Timeout to retrieve batch messages from the queue.
+        (Note: Applicable when all paths in `files` are local.)
```

Comment:
We perform a validation step for remote host paths, selecting the fsspec route if they exist, and subsequently confirming that they share the same protocol; otherwise, an error is raised.
It is implemented based on the description mentioned in this issue: https://github.com/nv-morpheus/Morpheus/issues/976